<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title></title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:sysadmin@kuat.pmacs.upenn.edu" />
</head>

<body style="background-color: white">



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a>
    <ul>
      <li><a href="#Requirements">Requirements</a></li>
    </ul>
  </li>
  <li><a href="#LICENSE">LICENSE</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<pre><code>   HiPR.sh - High-throughput Probabilistic inference of RNA secondary structures. </code></pre>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code>   HiPR.sh reads_file rates_file structure_file [-outDir OutputDir] [-locusName LocusName] [-n Niter] [-numCPU Ncpus]  [-rmin MinReadLength] [-rmax MaxReadLength]&quot;
   
   Mandatory arguments (need to be specified before optional arguments)
      reads_file -- File containing DMS-seq reads for RNA locus of interest (collapsed read format)
      rates_file -- File containing the initial estimates of per-nucleotide modification rates
      structure_file -- File containing the sequence and initial secondary structure of an RNA of interest

   Recognized optional command line arguments (need to be specified after mandatory arguments)
      -outDir &lt;string&gt;  -- Set name of output folder (default=HiPR_output)
      -locusName &lt;string&gt;  -- Set name of locus (default=UnnamedLocus)
      -n &lt;integer&gt; -- Set maximum number of MCMC iterations (default=100000)
      -numCPU &lt;integer&gt; -- Set number of CPUs to use (default=16)
      -rmin &lt;integer&gt; -- Set mininum read length (default=15)
      -rmax &lt;integer&gt; -- Set maximum read length (defauld=40)</code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>Estimate secondary structure and base pairing posteriors for a given RNA sequence based on the distribution of read fragments along the locus.</p>

<p>This program requires a file containing the sequence and initial secondary structure of an RNA of interest, a file containing DMS-seq reads, and a file containing the initial estimates of per-nucleotide modification rates. A Bayesian MCMC algorithm is then used to estimate the base pairing posterior that best fits the observed sequencing reads. The results and intermediate files are written to a directory (HiPR_output/ by default).</p>

<p>The output file HiPR_posterior.txt contains the base pairing posteriors at each nucleotide position, one entry per line.</p>

<p>The output file HiPR_structure.txt contains the consensus secondary structure.</p>

<h2 id="Requirements">Requirements</h2>

<p>Mandatory files must be specified before any other optional arguments and must exist, otherwise a error or usage message will be shown.</p>

<h1 id="LICENSE">LICENSE</h1>

<p>MIT License https://opensource.org/licenses/MIT</p>

<p>Copyright (c) 2016 University of Pennsylvania</p>

<p>Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the &quot;Software&quot;), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:</p>

<p>The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.</p>

<p>THE SOFTWARE IS PROVIDED &quot;AS IS&quot;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.</p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Pavel Kuksa &lt;pkuksa@upenn.edu&gt; Fan Li &lt;fanli.gcb@gmail.com&gt; Li-San Wang &lt;lswang@upenn.edu&gt;</p>


</body>

</html>


